File: website/models/gwas.py
#!/usr/bin/env python

from __future__ import print_function

from cre import CRE
from common.pg_gwas import PGgwas

class Gwas:
    def __init__(self, assembly, ps, cache):
        self.assembly = assembly
        self.cache = cache
        self.pgGwas = PGgwas(ps, assembly)

        self.studies = self.pgGwas.gwasStudies()
        self.byStudy = {r["value"] : r for r in self.studies}

    def checkStudy(self, gwas_study):
        return gwas_study in self.byStudy

    def numCresOverlap(self, gwas_study):
        return self.pgGwas.numCresOverlap(gwas_study)

    def totalLDblocks(self, gwas_study):
        return self.byStudy[gwas_study]["total_ldblocks"]

    def numLdBlocksOverlap(self, gwas_study):
        return self.pgGwas.numLdBlocksOverlap(gwas_study)

    def percCresEnhancer(self, gwas_study):
        return 0

    def percCresPromoter(self, gwas_study):
        return 0

    def allCellTypes(self, gwas_study):
        return self.pgGwas.gwasEnrichment(gwas_study)

    def cres(self, gwas_study, ct):
        cres, fieldsOut = self.pgGwas.gwasPercentActive(gwas_study, ct)

        # accession, snp, geneid, zscores
        totalActive = 0
        total = len(cres)
        activeCres = []

        def any_lambda(function, iterable):
            # http://stackoverflow.com/a/19868175
            return any(function(i) for i in iterable)

        for a in cres:
            if a.get("promoter zscore", 0) > 1.64 or a.get("enhancer zscore", 0) > 1.64 or a.get("dnase zscore", 0) > 1.64:
                totalActive += 1
                activeCres.append(a)

        hiddenFields = set(["promoter zscore", "enhancer zscore",
                            "dnase zscore"]) - set(fieldsOut)
        if hiddenFields:
            for r in cres:
                for f in hiddenFields:
                    r[f] = ''
        vcols = {}
        for f in ["promoter zscore", "enhancer zscore", "dnase zscore"]:
            vcols[f] = f not in hiddenFields
        return {"accessions" : activeCres,
                "vcols" : vcols}

File: graphql/src/resolvers/gwas.ts
import { GraphQLFieldResolver } from 'graphql';
import * as DbGwas from '../db/db_gwas';
import { loadCache, ccRECtspecificLoaders } from '../db/db_cache';
import { Assembly } from '../types';
import { UserInputError } from 'apollo-server-express';

export class Gwas {
    assembly: Assembly;
    studies: DbGwas.DBGwasStudy[];
    byStudy: Record<string, DbGwas.DBGwasStudy>;
    constructor(assembly) {
        this.assembly = assembly;
    }

    async awaitStudies() {
        const gwas_studies = await loadCache(this.assembly).gwas_studies();
        if (!this.studies) {
            this.studies = gwas_studies;
            this.byStudy = this.studies.reduce((obj, r) => {
                obj[r.name] = r;
                return obj;
            }, {});
        }
    }

    checkStudy = (study: string) => study in this.byStudy;

    numLdBlocksOverlap = gwas_study => DbGwas.numLdBlocksOverlap(this.assembly, gwas_study);
    numCresOverlap = gwas_study => DbGwas.numCresOverlap(this.assembly, gwas_study);
    allCellTypes = async gwas_study => {
        const datasets = await loadCache(this.assembly).datasets();
        const cts = await DbGwas.gwasEnrichment(this.assembly, gwas_study);
        const ret = cts ? cts.map(ct => ({ ...ct, ct: datasets.byCellTypeValue[ct.ct] })) : undefined;
        return ret;
    };

    async cres(gwas_study: string, ct: string | undefined) {
        const acache = loadCache(this.assembly);
        const ctmap = await acache.ctmap();
        const ctsTable = await acache.ctsTable();
        const celltype = ct !== 'none' ? ct : undefined;
        const cres = await DbGwas.gwasPercentActive(this.assembly, gwas_study, celltype, ctmap);
        let activeCres: DbGwas.gwascre[] = [];
        if (celltype) {
            const ctspecific = await ccRECtspecificLoaders[this.assembly].loadMany(
                cres.map(c => `${c.accession}::${celltype}`)
            );
            cres.forEach((cre, index) => {
                const cts = ctspecific[index];
                if (
                    (cts.h3k4me3_zscore || 0) > 1.64 ||
                    (cts.h3k27ac_zscore || 0) > 1.64 ||
                    (cts.dnase_zscore || 0) > 1.64
                ) {
                    activeCres.push(cre);
                }
            });
        } else {
            activeCres = cres;
        }

        // accession, snp, geneid, zscores
        return activeCres.map(c => ({ cRE: c, geneid: c.geneid, snps: c.snps }));
    }
}

export const resolve_gwas_studies: GraphQLFieldResolver<any, any> = source => {
    const g: Gwas = source.gwas_obj;
    return g.studies.map(study => ({
        study_name: study.name,
        gwas_obj: g,
        ...g.byStudy[study.name],
    }));
};

export const resolve_gwas_study: GraphQLFieldResolver<any, any> = (source, args) => {
    const g: Gwas = source.gwas_obj;
    const studyarg = args.study;
    if (!g.checkStudy(studyarg)) {
        throw new UserInputError('invalid gwas study');
    }
    return {
        study_name: studyarg,
        gwas_obj: g,
        ...g.byStudy[studyarg],
    };
};

export const resolve_gwas_study_numLdBlocksOverlap = async source => {
    const g: Gwas = source.gwas_obj;
    const study_name = source.study_name;
    return g.numLdBlocksOverlap(study_name);
};

export const resolve_gwas_study_numCresOverlap = async source => {
    const g: Gwas = source.gwas_obj;
    const study_name = source.study_name;
    return g.numCresOverlap(study_name);
};

export const resolve_gwas_study_allSNPs = async source => {
    const g: Gwas = source.gwas_obj;
    const study_name = source.study_name;
    const snps = await DbGwas.allSNPsInStudy(g.assembly, study_name, g);
    return snps;
};

export const resolve_gwas_study_topCellTypes = async source => {
    const g: Gwas = source.gwas_obj;
    const study_name = source.study_name;
    return g.allCellTypes(study_name);
};

export const resolve_gwas_study_cres: GraphQLFieldResolver<any, any> = async (source, args) => {
    const g: Gwas = source.gwas_obj;
    const study_name = source.study_name;
    const cellType = args.cellType;
    // TODO: check ct!
    return g.cres(study_name, cellType);
};

export const resolve_gwas_snps: GraphQLFieldResolver<any, any> = async (source, args, context, info) => {
    const g: Gwas = source.gwas_obj;
    const assembly = g.assembly;
    const search: string = args.search;
    return DbGwas.searchSNPs(assembly, search);
};

export const resolve_gwas: GraphQLFieldResolver<any, any> = (source, args, context, info) => {
    const assembly = args.assembly;
    const g = new Gwas(assembly);
    return g.awaitStudies().then(r => ({ gwas_obj: g }));
};
